### DISCUSSION TITLE
Failed to load external entity

### DISCUSSION THREAD

**Discussion by happypennygames** [w3c-cg/musicxml#445] [Q&A]
I am attempting to perform local validation of a musicxml file exported using musescore (uncompressed xml).  I'm using xmllint.js (https://github.com/kripken/xml.js/blob/master/xmllint.js) on a simulated client side using a python server.  The validation pipeline works great with other xsd/xml pairs, but I get this error when I substitute musicxml.xsd for the schema (has nothing to do with the xml file itself).

```
warning: failed to load external entity "http://www.musicxml.org/xsd/xlink.xsd"
file_0.xsd:26: element import: Schemas parser warning : Element '{http://www.w3.org/2001/XMLSchema}import': Failed to locate a schema at location 'http://www.musicxml.org/xsd/xml.xsd'. Skipping the import.
warning: failed to load external entity "http://www.musicxml.org/xsd/xlink.xsd"
file_0.xsd:27: element import: Schemas parser warning : Element '{http://www.w3.org/2001/XMLSchema}import': Failed to locate a schema at location 'http://www.musicxml.org/xsd/xlink.xsd'. Skipping the import.
file_0.xsd:2248: element attribute: Schemas parser error : attribute use (unknown), attribute 'ref': The QName value '{http://www.w3.org/XML/1998/namespace}lang' does not resolve to a(n) attribute declaration.
```
I copied the full zip of all the files into the same place as the musicxml.xsd fiile.  

The instructions in musicxml.xsd reference the catalog.xml file, but I do not reference it (I'm not sure how), so i don't know if that's the problem.  I'm also ok with manually pasting all the other xsd files into the musicxml.xsd file if there was  way to do that.

Modifying the musicxml.xsd file such that the schema location is just the file name, and running: 
```
xmllint --schema musicxml.xsd t1.musicxml --noout --nonet
```
actually does validate at the command line in a linux terminal, which implies the fault is in some interaction of the import statement with xmllint.js at this point.  Feel free to close the issue if no suggestions given this update.

**Comment by mdgood** [comment]
@happypennygames This page has instructions for using XML catalogs with lxml in Python. It sounds like this should work for your situation and is better than modifying the MusicXML schema files:

https://lxml.de/resolvers.html

Please let us know if this works for you.

**Comment by happypennygames** [comment]
Thanks for posting!  I did look at it as soon as you replied- unfortunately, couldn't make out how to use it.  I will try for a deeper look this weekend and let you know my specific questions.

**Comment by happypennygames** [comment]
The long and short of it- totally failed going that direction.  But that's ok, I just kept pushing forward, and I appreciated the pointer.  If you want to try out the application (free, no ads), you can upload simple uncompressed musicxml files (e.g. up to 8th notes) and try real time feedback with a keyboard- thanks again!    https://play.google.com/store/apps/details?id=com.happypennygames.aimusic.sightreader

**Comment by apacha** [comment]
Hi,
I've stumbled upon this problem too after upgrading to MusicXML Schema 4.0 DTD and I don't really see how the mentioned catalogues could solve the issue. I'm executing this very simple code, just trying to parse the schema:

```python
from lxml import etree

schema_file_name: str = 'utils/data_files/musicxml.xsd'
schema = etree.XMLSchema(file=schema_file_name)
```

and I already get the following error: 

```
Traceback (most recent call last):
  File "/Users/alex/musicxml_validation.py", line 5, in <module>
    schema = etree.XMLSchema(file=schema_file_name)
  File "src/lxml/xmlschema.pxi", line 88, in lxml.etree.XMLSchema.__init__
lxml.etree.XMLSchemaParseError: attribute use (unknown), attribute 'ref': The QName value '{http://www.w3.org/XML/1998/namespace}lang' does not resolve to a(n) attribute declaration., line 2247
```

which makes me question whether the schema can be used like that (by lxml) at all. For parsing the schema, there is no option for any catalog-resolver or the like.

Can you shed some light on this?

**Comment by mdgood** [comment]
@apacha Did you look at the linked article at https://lxml.de/resolvers.html? It explains the issue and provides links such as http://xmlsoft.org/catalog.html with information on how to set up catalogs for use with lxml.

**Comment by david-pace** [comment]
I just ran into the same issue when importing the MusicXML 4.0 XSD using the Eclipse Modeling Framework (EMF). If you're interested in some background information, you can read my [blog post](https://www.davidpace.de/generating-emf-models-from-xml-schema-definitions-xsds/).

The error messages indicate that the schema locations of the following XSD imports (lines 26 and 27 of `musicxml.xsd` in the current 4.0 release) can not be resolved:

```
<xs:import namespace="http://www.w3.org/XML/1998/namespace" schemaLocation="http://www.musicxml.org/xsd/xml.xsd"/>
<xs:import namespace="http://www.w3.org/1999/xlink" schemaLocation="http://www.musicxml.org/xsd/xlink.xsd"/>
```

When browsing the URLs directly in a browser the XSDs are reachable, but there is a HTTP -> HTTPS redirection (e.g. from `http://www.musicxml.org/xsd/xml.xsd` to `https://www.musicxml.org/xsd/xml.xsd`); also my browser (Firefox) indicates that the connection is not safe for some reason. Probably the HTTPS redirection, the SSL certificate or both cause problems while resolving the XSD imports.

My solution was to replace the URLs with the following ones:

```
https://www.musicxml.com/xsd/xml.xsd
https://www.musicxml.com/xsd/xlink.xsd
```

With those schema locations the MusicXML XSD could be loaded without any problems. Consequently one solution could be to use `https://` URLs in future versions.

**Comment by infojunkie** [comment]
I've [written a blog post](https://blog.karimratib.me/2020/11/17/validate-musicxml.html) about validating MusicXML using node.js - might be relevant here.

**Reply by apacha** [reply]
Thanks @infojunkie - I've resorted to the same patch as you gave, by replacing 

```xml
<xs:import namespace="http://www.w3.org/XML/1998/namespace" schemaLocation="http://www.musicxml.org/xsd/xml.xsd"/>
<xs:import namespace="http://www.w3.org/1999/xlink" schemaLocation="http://www.musicxml.org/xsd/xlink.xsd"/>
```
with
```xml
<xs:import namespace="http://www.w3.org/XML/1998/namespace" schemaLocation="xml.xsd"/>
<xs:import namespace="http://www.w3.org/1999/xlink" schemaLocation="xlink.xsd"/>
```

Given that all files are in the same directory anyway, I don't see the need for any resolvers. And yes, I looked at https://lxml.de/resolvers.html but it left me with more questions than it answered. Do I have to write my own resolver? Is there a default URL resolver somewhere (etree.Resolver)? Do I have to set some magic parameters somewhere? `etree.XMLSchema()` does not take any additional parameters so I don't know where to inject such a resolver in the first place, ... 

Bottom line: I'll stick to the simple solution of changing the schema. Thanks for your help!
